### DISCUSSION TITLE
[Feature] ML improvements

### DISCUSSION THREAD

**Discussion by mertalev** [immich-app/immich#2524] [Feature Request]
There are a number of improvements I'd like to see implemented for machine learning, and I'm open to implementing some of these myself. The most notable include:

- [X]  Hardware acceleration. This would allow users to utilize their GPU for faster inference.
- [X]  Model compilation. PyTorch models can be compiled into formats such as ONNX and TensorRT for faster inference and lower overhead. [Optimum](https://github.com/huggingface/optimum) notably makes this simple to apply to HF pipelines.
- [ ] Batching. Performing inference on one input at a time is less efficient than in batches, particularly for GPUs.
- [x] Model presets for different hardware levels. e.g. lower-end hardware could benefit from `buffalo_sc` over the larger `buffalo_l` facial recognition model, and might want to disable CLIP altogether. Likewise, higher-end hardware could allow for larger and more accurate models. 
- [x] Better object tagging. Described below.
- [X] Admin dashboard. It would be nice to expose ML settings in the UI to make it easier to adjust and more transparent. This would entail either communication between the main server and ML, and/or modifying a config file that both have access to. 

**Object Detection**

I believe the complaints around object tagging are largely caused by two factors:

1. Two different model tasks are used for object tagging: image classification and object detection. While both of these tag objects, the model objectives are different: the former tries to find the best object tag to describe an image, while the latter tries to find _every_ object in an image and describe each with a tag ([example](https://user-images.githubusercontent.com/61035602/208037368-5b9f01f7-afd9-46d8-bc80-271ccb5db7bb.png)). For users searching tagged images, image classification is the more relevant task as they're less interested in background objects.
2. While `resnet-50` is a decent (if older) model, `yolos-tiny` is subpar. Since object detection is also a more difficult task than image classification, this results in the object detection model being a net detriment to the tags.

In light of the above, I would suggest one of the following approaches:

1. Update the object detection model and set a minimum box size for objects to be added as tags, such that any box's area has to be a certain fraction of the image.
2. Use image classification tags alone for indexing purposes. Unless the box-level outputs are used (e.g. for the UI), the object detection model can be removed altogether (I've personally disabled it).

Thoughts? Does this seem like a good roadmap for ML?

### Platform

- [X] Server
- [x] Web
- [ ] Mobile

**Comment by alextran1502** [comment]
I think hardware acceleration is very much needed at the moment with the increase of machine learning jobs in the application and the availability of GPU in the server. This will help elevate some pain points in the system.

After CLIP is implemented, I don't see much need for object detection and image classification. Probably better to implement #2519 and make those keywords searchable in the Typesense server.

Let me know your thoughts

**Reply by mertalev** [reply]
That's an interesting model! But I think it would be too noisy for keywords, e.g. [here](https://user-images.githubusercontent.com/5921035/239678418-4ca182e5-fe3f-476f-9162-931e190b1e5d.png) most of the words don't make sense on their own ("school", "youtube", "vine", "Byron Galvez", etc.).

If tagging isn't too important, I would just scrap the object detection endpoint but keep image classification. A CLIP model trained on ImageNet would give higher quality tags. You can usually just take the top 5 they give without doing score filtering (or at least a much lower min score).

**Reply by MattPhilpot** [reply]
For those of us running k8s clusters without nvidia gpu's, how feasible is it to allow for more than one machine learning pod that can be assigned specific types of tasks? For example, my cluster has a good number of nodes, but each node isn't the highest spec'd (4 cores, 32GB memory). 

It would be nice to have the option to spin up five machine learning pods that each handle a specific job type.

**Reply by bo0tzz** [reply]
That should just work out of the box, albeit without assigning specific job types.

**Reply by MattPhilpot** [reply]
> That should just work out of the box, albeit without assigning specific job types.

How would I go about testing this? I don't see anything in the documentation about having multiple machine learning instances.

**Reply by bo0tzz** [reply]
You say you're using kubernetes - if you scale up your ML deployment, the service should automatically loadbalance between them and it should Just Work. You might need to scale up microservices as well to use the full potential, as there's a cap on the amount of tasks processed by one instance concurrently.

**Reply by MattPhilpot** [reply]
Good to know. Only reason I hadn't scaled up the two was because I wasn't sure how it'd be handled. I'll have to play around with i.

Thanks :D

**Reply by MattPhilpot** [reply]
Following up @bo0tzz. After adjusting the volumes a bit, it seems to be working as you described. Now have 4 replicas of both machine learning and microservices. Going to re-run a few jobs and see how fast they complete.

**Comment by ollioddi** [comment]
Do you have any AI accelerators i mind? Thinking of things lile Google Coral. My impression is that they still outperform most GPUs especially considering power draw.

**Reply by bo0tzz** [reply]
Coral doesn't support the framework (pytorch) that we use for ml. If you know of any accelerators that do (and that are reasonably available to consumers) we'd love to hear it.

**Reply by mertalev** [reply]
The initial implementation will likely just be for CUDA since it's the most straightforward, but I can also look into TPUs to see how much work it would be to support (Coral products are basically tiny TPUs).

**Reply by simone-viozzi** [reply]
You _can_ run pytorch on a TPU: https://github.com/pytorch/xla

With transformers this would require to use the model directly since `pipeline` does not support XLA: https://discuss.huggingface.co/t/how-do-i-do-inference-using-the-gpt-models-on-tpus/12644

TLDR: is doable but requires a lot of code and maintenance.

**Reply by ollioddi** [reply]
I am unfortunately not aware of any accelerator which runs pytorch. GPU's are very much accessible for many, but i do still think that a TPU or other ML specific devices would be an advantage to have, since they often fit in any systems via USB, M.2 and consume very little power in relation to computational power. But i do recognize the challenge, and it should not be a burden to implement. Guess i'll find a better use for my Coral :-)

**Comment by rhoot** [comment]
> For users searching tagged images, image classification is the more relevant task as they're less interested in background objects.

Personally I disagree, and it's one of the things I don't like with PhotoPrism. It's really bad at finding what you're looking for unless you happen to use the one exact tag it has assigned to that photo. I don't always remember what exactly was in a particular photo I may be looking for, and it's even less frequently that PhotoPrism agrees with me what the primary object is. As an example I have a photo of my car with my license plate visible, for the few times I need my license plate number. If I search for "license plate" in PhotoPrism it doesn't find that photo, since it was tagged with "car". Immich however finds it for both queries.

Not to mention the tags PhotoPrism assigns images are frequently either wrong or irrelevant. I have another photo of a box that was broken in transit before being delivered to me. PhotoPrism labelled it "Station House", and again I get no hits if I search for "box". Immich meanwhile manages to find it just fine.

Edit to clarify: I think image classification is useful when trying to categorize images and displaying that tag to the user. But for tags that don't show up in the UI, and that are just used as search terms, object classification is the winner in my book.

**Reply by bo0tzz** [reply]
Immich uses CLIP for search, which is more powerful than any of the tagging approaches.

**Reply by mertalev** [reply]
You're right, I'm not sure why I specified "for searching" when I had the tag UI in mind. As @bo0tzz  mentioned, since CLIP is already better for searching, that leaves object detection without a proper role: it's worse for searching and it makes the tags that show up in the UI more random.

**Comment by bo0tzz** [comment]
Closing as this has pretty much all been implemented
